### PR TITLE
New gir

### PR DESCRIPTION
**Pull request by GuillaumeGomez**

### CODE REVIEW

File: src/auto/assistant.rs [hunk 1]
```diff
@@ -233,9 +234,27 @@ impl<O: IsA<Assistant>> AssistantExt for O {
...
     }
 
-    //fn set_forward_page_func<'a, P: Into<Option<&'a /*Unimplemented*/AssistantPageFunc>>, Q: Into<Option</*Unimplemented*/Fundamental: Pointer>>>(&self, page_func: P, data: Q, destroy: /*Unknown conversion*//*Unimplemented*/DestroyNotify) {
-    //    unsafe { TODO: call ffi::gtk_assistant_set_forward_page_func() }
-    //}
+    fn set_forward_page_func<P: Fn(i32) -> i32 + 'static, Q: Into<Option<P>>>(&self, page_func: Q) {
+        let page_func = page_func.into();
+        let page_func_data: Box_<Box_<Option<P>>> = Box::new(Box::new(page_func.into()));
+        unsafe extern "C" fn page_func_func<P: Fn(i32) -> i32 + 'static>(current_page: libc::c_int, data: glib_ffi::gpointer) -> libc::c_int {
+            let callback: Box_<Box_<Option<P>>> = Box_::from_raw(data as *mut _);
```
Comment (sdroege) [inline src/auto/assistant.rs]: `from_raw()` will cause the closure to be freed at the end of the scope. Use simple references here instead, or `into_raw()` again at the end but that's less nice.

```rust
let callback = &*(data as *const Box_<Option<P>>);
```

File: src/auto/assistant.rs [hunk 2]
```diff
@@ -233,9 +234,27 @@ impl<O: IsA<Assistant>> AssistantExt for O {
...
+            let res = if let Some(ref callback) = **callback {
+                callback(current_page)
+            } else {
+                panic!("cannot get closure...")
+            };
+            res
+        }
+        let page_func = if page_func_data.is_some() { Some(page_func_func::<P> as _) } else { None };
+        unsafe extern "C" fn destroy_func<P: Fn(i32) -> i32 + 'static>(data: glib_ffi::gpointer) {
+         let _callback: Box_<Box_<Option<P>>> = Box_::from_raw(data as *mut _); }
```
Comment (sdroege) [inline src/auto/assistant.rs]: The indentation is broken here

**Comment by sdroege** [comment]
> The point of this PR is just to show, for now it isn't intended to be merged at all. :)

Sure, I'm just pointing out what is currently wrong or missing here :)

**Comment by GuillaumeGomez** [comment]
Oh my bad. Then yes, you're absolutely right! :)

File: src/auto/assistant.rs [hunk 3]
```diff
@@ -233,9 +234,27 @@ impl<O: IsA<Assistant>> AssistantExt for O {
         }
     }
 
-    //fn set_forward_page_func<'a, P: Into<Option<&'a /*Unimplemented*/AssistantPageFunc>>, Q: Into<Option</*Unimplemented*/Fundamental: Pointer>>>(&self, page_func: P, data: Q, destroy: /*Unknown conversion*//*Unimplemented*/DestroyNotify) {
-    //    unsafe { TODO: call ffi::gtk_assistant_set_forward_page_func() }
-    //}
+    fn set_forward_page_func<P: Fn(i32) -> i32 + 'static, Q: Into<Option<P>>>(&self, page_func: Q) {
+        let page_func = page_func.into();
+        let page_func_data: Box_<Box_<Option<P>>> = Box::new(Box::new(page_func.into()));
```
Comment (sdroege) [inline src/auto/assistant.rs]: As callback and destroy notify are working on the concrete type `P`, you only need to box once here. Boxing twice is only needed for trait objects

**Comment by sdroege** [comment]
This can be closed now I guess?

**Comment by GuillaumeGomez** [comment]
Indeed!

**Pull request closed without merging**